spirv_glsl.cpp

```diff
@@ -3583,12 +3583,22 @@ void CompilerGLSL::emit_extension_workarounds()
 		}
 		else if (shader_subgroup_supporter.is_feature_requested(Supp::SubgroupMemBarrier))
 		{
-			statement("#ifndef GL_KHR_shader_subgroup_basic");
-			statement("void subgroupMemoryBarrier() { groupMemoryBarrier(); }");
-			statement("void subgroupMemoryBarrierBuffer() { memoryBarrierBuffer(); }");
-			statement("void subgroupMemoryBarrierShared() { memoryBarrierShared(); }");
-			statement("void subgroupMemoryBarrierImage() { memoryBarrierImage(); }");
-			statement("#endif");
+			if (model == spv::ExecutionModelTessellationEvaluation || model == spv::ExecutionModelGLCompute)
+			{
+				statement("#ifndef GL_KHR_shader_subgroup_basic");
+				statement("void subgroupMemoryBarrier() { memoryBarrier(); }");
+				statement("void subgroupMemoryBarrierBuffer() { memoryBarrierBuffer(); }");
+				statement("void subgroupMemoryBarrierImage() { memoryBarrierImage(); }");
+				statement("#endif");
+			}
+			else
+			{
+				statement("#error Subgroup memory barriers can be emulated only in tesselation control and compute "
+				          "execution models.");
+			}
+
+			if (model == spv::ExecutionModelGLCompute)
+				statement("void subgroupMemoryBarrierShared() { memoryBarrierShared(); }");
```

Comment:
tessellation can use shared memory, so you can hoist it back.

Reply:
I've read GLSL spec and `memoryBarrierShared` is for compute only

Reply:
same for `groupMemoryBarrier`